conmon-rs/server/src/rpc.rs

```diff
@@ -48,29 +49,80 @@ impl conmon::Server for Server {
+                .await
+                .map_err(|e| IOError::new(ErrorKind::Other, format!("Error {}", e)))?;
 
-        let pid = pry_err!(pry!(fs::read_to_string(pidfile)).parse::<i32>());
-        let exit_paths = pry!(path_vec_from_text_list(pry!(req.get_exit_paths())));
-        let child = Child::new(id, pid, exit_paths);
-        pry_err!(child_reaper.watch_grandchild(&child));
+            debug!("Status for container ID {} is {}", id, status);
 
-        // register child with server
-        self.children.insert(child.id.clone(), child);
+            if let Some(console) = maybe_console {
+                console
+                    .wait_connected()
+                    .context("wait for console socket connection");
+            }
 
-        // TODO FIXME why convert?
-        results.get().init_response().set_container_pid(pid as u32);
-        Promise::ok(())
+            let pid = tokio::fs::read_to_string(pidfile)
+                .await?
+                .parse::<i32>()
+                .map_err(|e| IOError::new(ErrorKind::Other, format!("pid parse error {}", e)))?;
+
+            // register child with server
+            let child = Child::new(id, pid, exit_paths);
+            child_reaper.watch_grandchild(&child);
+            children.write().unwrap().insert(child.id.clone(), child);
+
+            // TODO FIXME why convert?
+            results.get().init_response().set_container_pid(pid as u32);
+            Ok(())
+        })
+    }
+
+    fn exec_sync_container(
+        &mut self,
+        params: conmon::ExecSyncContainerParams,
+        mut results: conmon::ExecSyncContainerResults,
+    ) -> Promise<(), capnp::Error> {
+        use std::io::{Error, ErrorKind};
+        let req = pry!(pry!(params.get()).get_request());
+        let id = pry!(req.get_id());
+        let timeout = req.get_timeout();
+        let command = self.generate_exec_sync_args(&params).unwrap();
+        let runtime = self.config.runtime().clone();
+        debug!(
+            "Got exec sync container request for id {} with timeout {} : {}",
+            id,
+            timeout,
+            command.join(" ")
+        );
+        let children = self.children.read().unwrap();
+        let child = children.get(id);
+        if child.is_none() {
+            let mut resp = results.get().init_response();
+            resp.set_exit_code(-1);
+            return Promise::ok(());
+        }
+        debug!("found child with id {}", child.unwrap().id);
+        let child_reaper = Arc::clone(self.reaper());
+        Promise::from_future(async move {
+            let (exit_status, stdout, stderr) = child_reaper
+                .exec_sync(&runtime, command, timeout)
+                .await
+                .map_err(|e| Error::new(ErrorKind::Other, format!("Error {}", e)))?;
+            let mut resp = results.get().init_response();
+            resp.set_exit_code(exit_status.code().unwrap());
+            resp.set_stdout(stdout.as_str());
```

Comment:
If `exec_sync` returned `Vec<u8>` this is where we could do `from_utf8_lossy()`.

Reply:
thanks. fixed :)